File: app/src/main/java/com/nrs/nsnik/notes/view/adapters/NotesAdapter.java
/*
 * Copyright (C) 2017 nsnikhil
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.nrs.nsnik.notes.view.adapters;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;
import com.jakewharton.rxbinding2.view.RxView;
import com.nrs.nsnik.notes.MyApplication;
import com.nrs.nsnik.notes.R;
import com.nrs.nsnik.notes.data.FolderEntity;
import com.nrs.nsnik.notes.data.NoteEntity;
import com.nrs.nsnik.notes.util.FileUtil;
import com.nrs.nsnik.notes.view.listeners.ItemTouchListener;
import com.nrs.nsnik.notes.view.listeners.NoteItemClickListener;

import java.io.File;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

/**
 * This adapter takes in a uri queries the uri and
 * registers for change in uri i.e. if any items in
 * that uri changes the adapter is notified and then
 * adapters clear the old list re draws the entire list
 * with the new data set
 */

public class NotesAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements ItemTouchListener {

    private static final int NOTES = 0, FOLDER = 1, HEADER = 2;
    private final Context mContext;
    private final LayoutInflater mLayoutInflater;
    private final RequestManager mRequestManager;
    private final FileUtil mFileUtil;
    private final CompositeDisposable mCompositeDisposable;
    private final File mRootFolder;
    private final NoteItemClickListener mNoteItemClickListener;
    private List<NoteEntity> mNotesList;
    private List<FolderEntity> mFolderList;

    /*
    TODO ENABLE SHOW TRANSITIONING OF LAYOUT CHANGES
     */

    /**
     * @param context    The context object
     * @param noteList   Note list
     * @param folderList Folder List
     */
    public NotesAdapter(Context context, @NonNull List<NoteEntity> noteList, @NonNull List<FolderEntity> folderList,
                        NoteItemClickListener noteItemClickListener) {
        mContext = context;
        mNotesList = noteList;
        mFolderList = folderList;
        mNoteItemClickListener = noteItemClickListener;

        mRequestManager = ((MyApplication) mContext.getApplicationContext()).getRequestManager();
        mFileUtil = ((MyApplication) mContext.getApplicationContext()).getFileUtil();
        mRootFolder = mFileUtil.getRootFolder();

        mLayoutInflater = LayoutInflater.from(mContext);
        mCompositeDisposable = new CompositeDisposable();
    }

    @Nullable
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case FOLDER:
                return new FolderViewHolder(mLayoutInflater.inflate(R.layout.single_folder_layout, parent, false));
            case NOTES:
                return new NoteViewHolder(mLayoutInflater.inflate(R.layout.single_note_layout, parent, false));
            case HEADER:
                return new HeaderViewHolder(mLayoutInflater.inflate(R.layout.single_list_header, parent, false));
        }
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int pos = mFolderList.size() + 2;
        int tempPOS = position - 1;
        switch (holder.getItemViewType()) {
            case FOLDER:
                bindFolderData(holder, tempPOS);
                break;
            case NOTES:
                pos = position - pos;
                bindNotesData(holder, pos);
                break;
            case HEADER:
                try {
                    StaggeredGridLayoutManager.LayoutParams layoutParams = (StaggeredGridLayoutManager.LayoutParams) holder.itemView.getLayoutParams();
                    layoutParams.setFullSpan(true);
                } catch (ClassCastException e) {
                    e.printStackTrace();
                }
                bindHeaderData(holder, position);
                break;
        }
    }

    /**
     * @param holder   HeaderViewHolder object
     * @param position Position of/in the list
     *                 <p>
     *                 this function binds the data for header view-holder type
     *                 it check if folder or note list is size is greater than 0
     *                 if yes than shows the appropriate headers otherwise
     *                 hides them
     */
    private void bindHeaderData(RecyclerView.ViewHolder holder, int position) {
        HeaderViewHolder headerViewHolder = (HeaderViewHolder) holder;
        if (headerViewHolder.mItemHeader != null) {
            if (position == 0) {
                if (mFolderList.size() > 0) {
                    headerViewHolder.mItemHeader.setVisibility(View.VISIBLE);
                    headerViewHolder.mItemHeader.setText(mContext.getResources().getString(R.string.headingFolder));
                } else {
                    headerViewHolder.mItemHeader.setVisibility(View.GONE);
                }
            } else {
                if (mNotesList.size() > 0) {
                    headerViewHolder.mItemHeader.setVisibility(View.VISIBLE);
                    headerViewHolder.mItemHeader.setText(mContext.getResources().getString(R.string.headingNotes));
                } else {
                    headerViewHolder.mItemHeader.setVisibility(View.GONE);
                }
            }
        }
    }

    /**
     * @param holder   FolderViewHolder object
     * @param position Position of/in the list
     *                 <p>
     *                 this function binds the data for FolderViewHolder type
     *                 it takes data from folder list and sets them on textview of
     *                 FolderViewHolder
     */
    private void bindFolderData(RecyclerView.ViewHolder holder, int position) {
        FolderViewHolder folderViewHolder = (FolderViewHolder) holder;
        if (folderViewHolder.mFolderNameText != null) {
            folderViewHolder.mFolderNameText.setText(mFolderList.get(position).getFolderName());
            folderViewHolder.mFolderNameText.setCompoundDrawableTintList(stateList(mFolderList.get(position).getColor()));
        }
    }

    /**
     * @param holder   NoteViewHolder object
     * @param position Position of/in the list
     *                 <p>
     *                 this function binds the data for NoteViewHolder type
     *                 it takes data from Notes list and sets title and
     *                 note content on textviews and others values.
     */
    private void bindNotesData(@NonNull RecyclerView.ViewHolder holder, int position) {
        final NoteViewHolder noteViewHolder = (NoteViewHolder) holder;
        NoteEntity object;
        try {
            object = mFileUtil.getNote(mNotesList.get(position).getFileName());

            //TITLE
            if (noteViewHolder.mNoteTitle != null) {
                if (object.getTitle() != null && !object.getTitle().isEmpty()) {
                    noteViewHolder.mNoteTitle.setVisibility(View.VISIBLE);
                    noteViewHolder.mNoteTitle.setText(object.getTitle());
                    noteViewHolder.mNoteTitle.setTextColor(Color.parseColor(object.getColor()));
                } else {
                    noteViewHolder.mNoteTitle.setVisibility(View.GONE);
                }
            }

            //CONTENT
            if (noteViewHolder.mNoteContent != null) {
                if (object.getNoteContent() != null && !object.getNoteContent().isEmpty()) {
                    noteViewHolder.mNoteContent.setVisibility(View.VISIBLE);
                    noteViewHolder.mNoteContent.setText(object.getNoteContent());
                } else {
                    noteViewHolder.mNoteContent.setVisibility(View.GONE);
                }

            }

            //IMAGES
            if (noteViewHolder.mNoteImage != null) {
                if (object.getImageList() != null && object.getImageList().size() > 0) {
                    noteViewHolder.mNoteImage.setVisibility(View.VISIBLE);
                    mRequestManager.load(new File(mRootFolder, object.getImageList().get(0))).into(noteViewHolder.mNoteImage);
                } else {
                    noteViewHolder.mNoteImage.setVisibility(View.GONE);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    @Override
    public int getItemCount() {
        return mNotesList != null && mFolderList != null ? mNotesList.size() + mFolderList.size() + 2 : 2;
    }

    //TODO CHANGE NOTIFY-DATA-SET-CHANGE WITH DIFF UTIL
    public void updateNotesList(@NonNull List<NoteEntity> noteList) {
        mNotesList = noteList;
        notifyDataSetChanged();
    }

    //TODO CHANGE NOTIFY-DATA-SET-CHANGE WITH DIFF UTIL
    public void updateFolderList(@NonNull List<FolderEntity> folderList) {
        mFolderList = folderList;
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0 || position == mFolderList.size() + 1) {
            return HEADER;
        } else if (position > 0 && position <= mFolderList.size()) {
            return FOLDER;
        } else if (position > mFolderList.size() + 1 && position < mNotesList.size()) {
            return NOTES;
        }
        return super.getItemViewType(position);
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
        if (viewHolder.getItemViewType() == target.getItemViewType()) {
            notifyItemMoved(fromPosition, toPosition);
        }
    }

    //TODO
    @Override
    public void onItemMoved(int fromPosition, int toPosition, RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        int startPos = -100;
        if (viewHolder.getItemViewType() == NOTES) {
            startPos = mFolderList.size() + 2;
        }
        if (viewHolder.getItemViewType() == FOLDER) {
            startPos = 1;
        }
        int fromPos = fromPosition - startPos;
        int toPos = toPosition - startPos;
    }

    //TODO
    @Override
    public void onItemDismiss(int position) {
        //notifyItemRemoved(position);
    }

    @NonNull
    private ColorStateList stateList(String colorString) {
        int[][] states = new int[][]{
                new int[]{android.R.attr.state_enabled},
                new int[]{-android.R.attr.state_enabled},
                new int[]{-android.R.attr.state_checked},
                new int[]{android.R.attr.state_pressed}
        };
        int color = Color.parseColor(colorString);
        int[] colors = new int[]{color, color, color, color};
        return new ColorStateList(states, colors);
    }

    private void cleanUp() {
        mCompositeDisposable.dispose();
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        cleanUp();
        super.onDetachedFromRecyclerView(recyclerView);
    }

    class NoteViewHolder extends RecyclerView.ViewHolder {
        @Nullable
        @BindView(R.id.singleNoteTitle)
        TextView mNoteTitle;

        @Nullable
        @BindView(R.id.singleNoteContent)
        TextView mNoteContent;

        @Nullable
        @BindView(R.id.singleNoteImage)
        ImageView mNoteImage;

        NoteViewHolder(@NonNull final View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            mCompositeDisposable.add(RxView.clicks(itemView).subscribe(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    mNoteItemClickListener.onClick(getAdapterPosition(), getItemViewType());
                }
            }, throwable -> Timber.d(throwable.getMessage())));

            mCompositeDisposable.add(RxView.longClicks(itemView).subscribe(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    mNoteItemClickListener.onLongClick(getAdapterPosition(), getItemViewType());
                }
            }, throwable -> Timber.d(throwable.getMessage())));
        }
    }


    class FolderViewHolder extends RecyclerView.ViewHolder {
        @Nullable
        @BindView(R.id.singleFolderName)
        TextView mFolderNameText;

        FolderViewHolder(@NonNull final View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            mCompositeDisposable.add(RxView.clicks(itemView).subscribe(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    mNoteItemClickListener.onClick(getAdapterPosition(), getItemViewType());
                }
            }));

            mCompositeDisposable.add(RxView.longClicks(itemView).subscribe(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    mNoteItemClickListener.onLongClick(getAdapterPosition(), getItemViewType());
                }
            }, throwable -> Timber.d(throwable.getMessage())));
        }
    }

    class HeaderViewHolder extends RecyclerView.ViewHolder {
        @Nullable
        @BindView(R.id.itemHeader)
        TextView mItemHeader;

        HeaderViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/nrs/nsnik/notes/view/adapters/NotesAdapter.kt
/*
 *     Credit Card Security V1  Copyright (C) 2018  sid-sun
 *     This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
 *     This is free software, and you are welcome to redistribute it
 *     under certain conditions; type `show c' for details.
 *
 * The hypothetical commands `show w' and `show c' should show the appropriate
 * parts of the General Public License.  Of course, your program's commands
 * might be different; for a GUI interface, you would use an "about box".
 *
 *   You should also get your employer (if you work as a programmer) or school,
 * if any, to sign a "copyright disclaimer" for the program, if necessary.
 * For more information on this, and how to apply and follow the GNU GPL, see
 * <http://www.gnu.org/licenses/>.
 *
 *   The GNU General Public License does not permit incorporating your program
 * into proprietary programs.  If your program is a subroutine library, you
 * may consider it more useful to permit linking proprietary applications with
 * the library.  If this is what you want to do, use the GNU Lesser General
 * Public License instead of this License.  But first, please read
 * <http://www.gnu.org/philosophy/why-not-lgpl.html>.
 */

package com.nrs.nsnik.notes.view.adapters

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.RequestManager
import com.jakewharton.rxbinding2.view.RxView
import com.nrs.nsnik.notes.MyApplication
import com.nrs.nsnik.notes.R
import com.nrs.nsnik.notes.data.FolderEntity
import com.nrs.nsnik.notes.data.NoteEntity
import com.nrs.nsnik.notes.util.AppUtil
import com.nrs.nsnik.notes.util.FileUtil
import com.nrs.nsnik.notes.view.listeners.ItemTouchListener
import com.nrs.nsnik.notes.view.listeners.NoteItemClickListener
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.item_indicator.view.*
import kotlinx.android.synthetic.main.single_folder_layout.view.*
import kotlinx.android.synthetic.main.single_list_header.view.*
import kotlinx.android.synthetic.main.single_note_layout.view.*
import timber.log.Timber
import java.io.File


/**
 * This adapter takes in a uri queries the uri and
 * registers for change in uri i.e. if any items in
 * that uri changes the adapter is notified and then
 * adapters clear the old list re draws the entire list
 * with the new data set
 */

class NotesAdapter(private val mContext: Context,
                   private var mNotesList: List<NoteEntity>,
                   private var mFolderList: List<FolderEntity>,
                   private val mNoteItemClickListener: NoteItemClickListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), ItemTouchListener {

    companion object {
        private const val NOTES = 0
        private const val FOLDER = 1
        private const val HEADER = 2
    }

    private val mLayoutInflater: LayoutInflater
    private val mRequestManager: RequestManager?
    private val mFileUtil: FileUtil?
    private val mCompositeDisposable: CompositeDisposable
    private val mRootFolder: File

    init {

        mRequestManager = (mContext.applicationContext as MyApplication).requestManager
        mFileUtil = (mContext.applicationContext as MyApplication).fileUtil
        mRootFolder = mFileUtil.rootFolder

        mLayoutInflater = LayoutInflater.from(mContext)
        mCompositeDisposable = CompositeDisposable()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            FOLDER -> FolderViewHolder(mLayoutInflater.inflate(R.layout.single_folder_layout, parent, false))
            NOTES -> NoteViewHolder(mLayoutInflater.inflate(R.layout.single_note_layout, parent, false))
            HEADER -> HeaderViewHolder(mLayoutInflater.inflate(R.layout.single_list_header, parent, false))
            else -> NoteViewHolder(mLayoutInflater.inflate(R.layout.single_note_layout, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        var pos = mFolderList.size + 2
        val tempPOS = position - 1
        when (holder.itemViewType) {
            FOLDER -> bindFolderData(holder, tempPOS)
            NOTES -> {
                pos = position - pos
                bindNotesData(holder, pos)
            }
            HEADER -> {
                try {
                    val layoutParams = holder.itemView.layoutParams as StaggeredGridLayoutManager.LayoutParams
                    layoutParams.isFullSpan = true
                } catch (e: ClassCastException) {
                    e.printStackTrace()
                }
                bindHeaderData(holder, position)
            }
        }
    }

    override fun getItemCount(): Int {
        return mNotesList.size + mFolderList.size + 2
    }

    private fun bindHeaderData(holder: RecyclerView.ViewHolder, position: Int) {
        val headerViewHolder = holder as HeaderViewHolder
        if (position == 0) {
            headerViewHolder.itemHeader.visibility = if (mFolderList.isNotEmpty()) View.VISIBLE else View.GONE
            if (headerViewHolder.itemHeader.isVisible) headerViewHolder.itemHeader.text = mContext.resources.getString(R.string.headingFolder)
        } else {
            headerViewHolder.itemHeader.visibility = if (mNotesList.isNotEmpty()) View.VISIBLE else View.GONE
            if (headerViewHolder.itemHeader.isVisible) headerViewHolder.itemHeader.text = mContext.resources.getString(R.string.headingNotes)
        }
    }

    private fun bindFolderData(holder: RecyclerView.ViewHolder, position: Int) {
        val folderViewHolder = holder as FolderViewHolder
        val folderEntity: FolderEntity = mFolderList[position]
        folderViewHolder.folderName.text = folderEntity.folderName
        folderViewHolder.folderName.compoundDrawableTintList = AppUtil.stateList(folderEntity.color)
        changeVisibility(folderViewHolder.isPinned, folderEntity.pinned)
        changeVisibility(folderViewHolder.isLocked, folderEntity.locked)
    }

    private fun bindNotesData(holder: RecyclerView.ViewHolder, position: Int) {
        val noteViewHolder = holder as NoteViewHolder
        val noteEntity: NoteEntity
        try {
            noteEntity = mFileUtil!!.getNote(mNotesList[position].fileName!!)

            //TITLE
            if (noteEntity.title != null && !noteEntity.title!!.isEmpty()) {
                noteViewHolder.noteTitle.visibility = View.VISIBLE
                noteViewHolder.noteTitle.text = noteEntity.title
                noteViewHolder.noteTitle.setTextColor(Color.parseColor(noteEntity.color))
            } else {
                noteViewHolder.noteTitle.visibility = View.GONE
            }

            if (noteEntity.locked == 0) {
                if (noteEntity.noteContent != null && !noteEntity.noteContent!!.isEmpty()) {
                    noteViewHolder.noteContent.visibility = View.VISIBLE
                    noteViewHolder.noteContent.text = noteEntity.noteContent
                    noteViewHolder.noteContent.setTextColor(ContextCompat.getColor(mContext, R.color.colorAccent))
                    noteViewHolder.noteContent.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, 0, 0)
                    noteViewHolder.noteContent.compoundDrawablePadding = 0
                } else {
                    noteViewHolder.noteContent.visibility = View.GONE
                }

                if (noteEntity.imageList != null && noteEntity.imageList!!.isNotEmpty()) {
                    noteViewHolder.noteImage.visibility = View.VISIBLE
                    mRequestManager?.load(File(mRootFolder, noteEntity.imageList!![0]))?.into(noteViewHolder.noteImage)
                } else {
                    noteViewHolder.noteImage.visibility = View.GONE
                }

            } else {
                noteViewHolder.noteContent.text = mContext.resources?.getString(R.string.hidden)
                noteViewHolder.noteImage.visibility = View.GONE
                noteViewHolder.noteContent.setTextColor(ContextCompat.getColor(mContext, R.color.grey))
                noteViewHolder.noteContent.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_lock_black_48px, 0, 0, 0)
                noteViewHolder.noteContent.compoundDrawablePadding = 4
            }

            changeVisibility(noteViewHolder.isPinned, noteEntity.pinned)

        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun changeVisibility(imageView: ImageView, value: Int) {
        imageView.visibility = if (value == 1) View.VISIBLE else View.GONE
    }

    //TODO CHANGE NOTIFY-DATA-SET-CHANGE WITH DIFF UTIL
    fun updateNotesList(noteList: List<NoteEntity>) {
        mNotesList = noteList
        notifyDataSetChanged()
    }

    //TODO CHANGE NOTIFY-DATA-SET-CHANGE WITH DIFF UTIL
    fun updateFolderList(folderList: List<FolderEntity>) {
        mFolderList = folderList
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0 || position == mFolderList.size + 1) HEADER
        else if (position > 0 && position <= mFolderList.size) FOLDER
        else if (position > mFolderList.size + 1 && position < mNotesList.size) NOTES
        else super.getItemViewType(position)
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder) {
        if (viewHolder.itemViewType == target.itemViewType) {
            notifyItemMoved(fromPosition, toPosition)
        }
    }

    //TODO
    override fun onItemMoved(fromPosition: Int, toPosition: Int, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        var startPos = -100
        if (viewHolder.itemViewType == NOTES) {
            startPos = mFolderList.size + 2
        }
        if (viewHolder.itemViewType == FOLDER) {
            startPos = 1
        }
        val fromPos = fromPosition - startPos
        val toPos = toPosition - startPos
    }

    //TODO
    override fun onItemDismiss(position: Int) {
        //notifyItemRemoved(position);
    }

    private fun cleanUp() {
        mCompositeDisposable.dispose()
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        cleanUp()
        super.onDetachedFromRecyclerView(recyclerView)
    }

    internal inner class NoteViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val noteTitle: TextView = itemView.singleNoteTitle
        val noteContent: TextView = itemView.singleNoteContent
        val noteImage: ImageView = itemView.singleNoteImage
        val isPinned: ImageView = itemView.itemIndicatorPin
        val isLocked: ImageView = itemView.itemIndicatorLock

        init {

            mCompositeDisposable.addAll(
                    RxView.clicks(itemView).subscribe({
                        if (adapterPosition != RecyclerView.NO_POSITION) {
                            mNoteItemClickListener.onClick(adapterPosition, itemViewType)
                        }
                    }, { throwable -> Timber.d(throwable.message) }),

                    RxView.longClicks(itemView).subscribe({
                        if (adapterPosition != RecyclerView.NO_POSITION) {
                            mNoteItemClickListener.onLongClick(adapterPosition, itemViewType, itemView)
                        }
                    }, { throwable -> Timber.d(throwable.message) })
            )
        }
    }


    internal inner class FolderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val folderName: TextView = itemView.singleFolderName
        val isPinned: ImageView = itemView.itemIndicatorPin
        val isLocked: ImageView = itemView.itemIndicatorLock

        init {
            mCompositeDisposable.addAll(
                    RxView.clicks(itemView).subscribe {
                        if (adapterPosition != RecyclerView.NO_POSITION) {
                            mNoteItemClickListener.onClick(adapterPosition, itemViewType)
                        }
                    },
                    RxView.longClicks(itemView).subscribe({
                        if (adapterPosition != RecyclerView.NO_POSITION) {
                            mNoteItemClickListener.onLongClick(adapterPosition, itemViewType, itemView)
                        }
                    }, { throwable -> Timber.d(throwable.message) })

            )
        }
    }

    internal inner class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val itemHeader: TextView = itemView.itemHeader

        init {
        }
    }
}